### PR TITLE
Add options for alternate plugin manifest sources (local/remote)

### PR DESCRIPTION
**Pull request by karthik2804**
Added options to include alternate sources for manifest - Local and Remote.

Signed-off-by: karthik Ganeshram <karthik.ganeshram@fermyon.com>

### CODE REVIEW

File: crates/plugins/src/install.rs [hunk 1]
```diff
@@ -16,28 +18,95 @@ const PLUGIN_MANIFESTS_DIRECTORY_NAME: &str = "manifests";
 
 pub struct PluginInstaller {
     name: String,
+    local_manifest_src: Option<String>,
+    remote_manifest_src: Option<Url>,
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: Since these are mutually exclusive, should they be wrapped in an enum (maybe with a third option of plugins repo (default))

File: src/commands/plugins.rs [hunk 2]
```diff
@@ -35,19 +35,58 @@ impl PluginCommands {
...
+    #[clap(
+        name = "REMOTE_PLUGIN_MANIFEST",
+        short = 'u',
+        long = "url",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST"
+    )]
+    pub remote_manifest_src: Option<Url>,
+    /// skips prompt to accept the installation of the plugin.
+    #[clap(short = 'y', long = "yes", takes_value = false)]
+    pub skip_accept_install: bool,
```
Comment (kate-goldenring) [inline src/commands/plugins.rs]: hmm I think we could name this differently. Maybe, `yes_to_all` or `assume_yes`. This field right now i see as the equivalent of `-y` in `apt` in that it is yes to all prompts. Right now we happen to only have one but maybe we should give it a general name, assuming there may be more in the future.

File: src/commands/plugins.rs [hunk 3]
```diff
@@ -35,19 +35,58 @@ impl PluginCommands {
...
+    /// specify particular version of plugin to install from centralized repository
+    #[clap(
+        long = "version",
+        short = 'v',
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        requires("PLUGIN_NAME")
+    )]
+    pub version: Option<String>,
```
Comment (kate-goldenring) [inline src/commands/plugins.rs]: I wonder if we can use clap's [`value_parser`](https://docs.rs/clap/latest/clap/macro.value_parser.html) here to immediately parse that this can be parsed as SemVer by the [`semver` crate](https://docs.rs/semver/1.0.13/semver/#example)

File: src/commands/plugins.rs [hunk 4]
```diff
@@ -35,19 +35,58 @@ impl PluginCommands {
...
 }
 
 impl Install {
     pub async fn run(self) -> Result<()> {
-        println!("The name of the plugin being installed {:?}", self.name);
-        PluginInstaller::new(&self.name, SPIN_PLUGINS_REPO, get_spin_plugins_directory()?)?
-            .install()
-            .await?;
+        println!("Attempting to install plugin: {:?}", self.name);
```
Comment (kate-goldenring) [inline src/commands/plugins.rs:81]: Maybe we are a little bit more confident and say "Installing plugin...". Or we say "Looking up plugin {} for installation"

File: src/commands/plugins.rs [hunk 5]
```diff
@@ -35,19 +35,58 @@ impl PluginCommands {
...
+        PluginInstaller::new(
+            match &self.name {
+                Some(name) => name,
+                _ => "",
```
Comment (kate-goldenring) [inline src/commands/plugins.rs]: When will there be no name here? This should be required.

File: src/commands/plugins.rs [hunk 6]
```diff
@@ -35,19 +35,58 @@ impl PluginCommands {
...
-    // If present, updates existing plugins instead of skipping.
-    // TODO: think about how to handle breaking changes
-    // #[structopt(long = "update")]
-    // pub update: bool,
+    #[clap(
+        name = "PLUGIN_NAME",
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST"
+    )]
+    pub name: Option<String>,
```
Comment (kate-goldenring) [inline src/commands/plugins.rs:45]: Name should be required. Remove option?

File: crates/plugins/Cargo.toml [hunk 7]
```diff
@@ -9,6 +9,7 @@ edition = "2021"
 anyhow = "1.0"
 bytes = "1.1"
 flate2 = "1.0"
+git2 = "0.15"
```
Comment (kate-goldenring) [inline crates/plugins/Cargo.toml]: Spin Templates do not use this helper crate rather build commands (see `templates/src/source.rs`). We may want to do the same instead of using a crate if possible

File: crates/plugins/src/install.rs [hunk 8]
```diff
@@ -16,28 +18,95 @@ const PLUGIN_MANIFESTS_DIRECTORY_NAME: &str = "manifests";
...
     }
+
     pub async fn install(&self) -> Result<()> {
         // TODO: Potentially handle errors to give useful error messages
+        let manifest_url = match self.remote_manifest_src.to_owned() {
+            Some(remote_manifest_url) => remote_manifest_url,
+            _ => self.url.to_owned(),
+        };
+
+        let plugin_manifest: Plugin = match self {
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: This is cool matching but i think this should be explicitly mutual exclusive via an enum. For example, what if they provide a remote url and local path? Only the remote will be used when this is input error. We should probably immediately catch this with clap

File: crates/plugins/src/install.rs [hunk 9]
```diff
@@ -16,28 +18,95 @@ const PLUGIN_MANIFESTS_DIRECTORY_NAME: &str = "manifests";
...
+            // Remote manifest source is provided
+            PluginInstaller {
+                remote_manifest_src: Some(remote_manifest_url),
+                ..
+            } => {
+                println!("installing from remote source");
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: These printlns are misleading. We are not installing yet. More like "Getting plugin manifest from ..."

File: crates/plugins/src/install.rs [hunk 10]
```diff
@@ -16,28 +18,95 @@ const PLUGIN_MANIFESTS_DIRECTORY_NAME: &str = "manifests";
...
+                println!("installing from local source");
+                let file = File::open(local_manifest_file)?;
+                serde_json::from_reader(file)?
+            }
+            _ => {
+                println!(
+                    "installing from central source, {} |||   {:#?}",
+                    self.url, self.plugins_dir
+                );
+                if !self.plugins_dir.join(".plugins").exists() {
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: Are we assuming the plugin repo is cloned if this dir exists? Might be better to explicitly check for the repo one level lower `.plugins/spin-plugins` in case the folder is created but not filled

File: crates/plugins/src/install.rs [hunk 11]
```diff
@@ -16,28 +18,95 @@ const PLUGIN_MANIFESTS_DIRECTORY_NAME: &str = "manifests";
...
+                    self.get_latest_plugin_repo()?;
+                } else {
+                    self.update_plugins_repository()?;
+                }
+                let file = File::open(
+                    &self
+                        .plugins_dir
+                        .join(".plugins/plugins")
+                        .join(get_manifest_file_name(&self.name)),
+                )?;
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs:102]: We might want to handle this error to be more descriptive, since this is where an error would occur if they try to install a plugin that does not exist in `spin-plugins`. maybe map the error.

File: crates/plugins/src/install.rs [hunk 12]
```diff
@@ -16,28 +18,95 @@ const PLUGIN_MANIFESTS_DIRECTORY_NAME: &str = "manifests";
...
+
+        // If local file provided, use the manifest
+        // else, grab manifest from remote source.
+        // let plugin: Plugin = match self.local_manifest_src.to_owned() {
+        //     Some(local_manifest_file) => {
+        //         let file = File::open(local_manifest_file)?;
+        //         serde_json::from_reader(file)?
+        //     }
+        //     _ => reqwest::get(manifest_url).await?.json::<Plugin>().await?,
+        // };
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: Is this commented section needed or can it be removed

File: crates/plugins/src/install.rs [hunk 13]
```diff
@@ -48,38 +117,55 @@ impl PluginInstaller {
...
-            self.url.to_owned(),
-            &target_url,
-        )?
-        .run()?
-        {
-            // User has requested to not install package, returning early
-            println!("Plugin {} will not be installed", plugin.name);
-            return Ok(());
+
+        // Ask for user confirmation if not overridden wiht CLI option
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: typo: SpellChecker is a great vscode plugin for these

File: crates/plugins/src/install.rs [hunk 14]
```diff
@@ -48,38 +117,55 @@ impl PluginInstaller {
...
         // Save manifest to installed plugins directory
-        self.add_to_manifest_dir(&plugin)?;
+        self.add_to_manifest_dir(&plugin_manifest)?;
+        log::info!("Plugin: {} installed successfully", self.name);
         Ok(())
     }
 
-    fn untar_plugin(&self, plugin_file_name: &PathBuf) -> Result<()> {
+    fn get_latest_plugin_repo(&self) -> Result<()> {
+        Repository::clone(&self.url.to_string(), self.plugins_dir.join(".plugins"))?;
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: This is where we may want to use the approach Spin templates is using instead of an extra crate

Comment (kate-goldenring) [reply crates/plugins/src/install.rs]: We could move this functionality into a git crate that templates coudl reference too

File: crates/plugins/src/install.rs [hunk 15]
```diff
@@ -48,38 +117,55 @@ impl PluginInstaller {
...
+        Ok(())
+    }
+
+    fn update_plugins_repository(&self) -> Result<(), git2::Error> {
+        let remote_name = "origin";
+        let remote_branch = "main";
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: These should be constants

File: crates/plugins/src/install.rs [hunk 16]
```diff
@@ -48,38 +117,55 @@ impl PluginInstaller {
...
+        let repo = Repository::open(self.plugins_dir.join(".plugins"))?;
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: This `.plugins` dir should be a constant

File: crates/plugins/src/install.rs [hunk 17]
```diff
@@ -89,7 +175,7 @@ impl PluginInstaller {
         // TODO: this is unix only. Look into whether permissions are preserved
         archive.set_preserve_permissions(true);
         // Create subdirectory in plugins directory for this plugin
-        let plugin_sub_dir = self.plugins_dir.join(&self.name);
+        let plugin_sub_dir = self.plugins_dir.join(plugin_name);
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs:156]: Why is this changed instead of `self.name`? `plugin_manifest.name` == `plugin $name` == `$name.json` right?

File: crates/plugins/src/install.rs [hunk 18]
```diff
@@ -147,3 +233,170 @@ fn file_digest_string(path: impl AsRef<Path>) -> Result<String> {
...
+    repo: &'a git2::Repository,
+    refs: &[&str],
+    remote: &'a mut git2::Remote,
+) -> Result<git2::AnnotatedCommit<'a>, git2::Error> {
+    let mut cb = git2::RemoteCallbacks::new();
+
+    // Print out our transfer progress.
+    cb.transfer_progress(|stats| {
+        if stats.received_objects() == stats.total_objects() {
+            print!(
```
Comment (kate-goldenring) [inline crates/plugins/src/install.rs]: i think many of these prints should be tracing but we can do a cleanup later before PR if this is helpful for debugging

**Review by kate-goldenring — COMMENTED** [review]
Thanks for adding all of this. I think the git repository fetching utils can be moved into their own file/module and ideally not use the git2 crate and instead take an approach similar to templates... they could maybe even share the same utils

File: crates/plugins/src/install.rs (hunk 1 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: I think that would be a good idea.

File: src/commands/plugins.rs (hunk 2 above)
Comment (karthik2804) [reply src/commands/plugins.rs]: `yes_to_all` seems better. I will make this change.

File: src/commands/plugins.rs (hunk 3 above)
Comment (karthik2804) [reply src/commands/plugins.rs]: I will look into this and if possible make the change.

File: src/commands/plugins.rs (hunk 5 above)
Comment (karthik2804) [reply src/commands/plugins.rs]: It feels a little less ergonomic to specify the name when installing from either a local/remote manifest.

`spin plugin install -f <path_to_plugin>` seems more ergonomic and also since we are pointing at the manifest, we will be able to obtain the plugin name from the manifest. The plugin name is only explicitly required when we fetch from the centralized repository to identify which particular manifest needs to be opened.

File: src/commands/plugins.rs (hunk 6 above)
Comment (karthik2804) [reply src/commands/plugins.rs:45]: See above.

File: crates/plugins/src/install.rs (hunk 8 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: When both a remote URL and local path are provided, Clap flags that as an input error as they are mutually exclusive arguments as indicated by the `conflicts _with` flag while building the parser.

File: crates/plugins/src/install.rs (hunk 10 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: Good point but I think it would be better to checkfor `.plugins/.git`

File: crates/plugins/src/install.rs (hunk 11 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs:102]: Agree

File: crates/plugins/src/install.rs (hunk 12 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: It can be removed

File: crates/plugins/src/install.rs (hunk 15 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: agree

File: crates/plugins/src/install.rs (hunk 16 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: agree

File: crates/plugins/src/install.rs (hunk 17 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs:156]: The name is not required when installing from local and remote manifests as described above.

File: crates/plugins/src/install.rs (hunk 14 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: Sounds like a good idea

File: crates/plugins/src/install.rs (hunk 9 above)
Comment (karthik2804) [reply crates/plugins/src/install.rs]: I will go through the print statements once more to make them more descriptive of the actual activity. I have been using them to debug.
